src/protections/MentionSpam.ts

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2024 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Protection } from "./IProtection";
+import { Mjolnir } from "../Mjolnir";
+import { LogLevel, Permalinks, UserID } from "@vector-im/matrix-bot-sdk";
+import { NumberProtectionSetting } from "./ProtectionSettings";
+
+export const DEFAULT_MAX_MENTIONS = 10;
+
+export class MentionSpam extends Protection {
+
+    settings = {
+        maxMentions: new NumberProtectionSetting(DEFAULT_MAX_MENTIONS, 1),
+    };
+
+    constructor() {
+        super();
+    }
+
+    public get name(): string {
+        return 'MentionSpam';
+    }
+    public get description(): string {
+        return "If a user posts many mentions, that message is redacted. No bans are issued.";
+    }
+
+    public checkMentions(body: unknown|undefined, htmlBody: unknown|undefined, mentionArray: unknown|undefined): boolean {
+        const max = this.settings.maxMentions.value;
+        if (Array.isArray(mentionArray) && mentionArray.length > max) {
+            return true;
+        }
+        if (typeof body === "string" && body.split('@').length - 1 > max) {
+            return true;
+        }
+        if (typeof htmlBody === "string" && htmlBody.split('%40').length - 1 > max) {
```

Comment:
I don't think all clients urlencode mxids properly in the anchor hrefs. So just a heads up that this isn't perfect

Reply:
(famously element didnt in the past. Not sure if this has been fixed. I just remember other client devs implementing workarounds for this case)

Reply:
why even have standards :fire: 

thanks, I'll address this one when I'm back if someone doesn't get to it first.